### ISSUE TITLE
Transforms: Config from query results, add support for more thresholds

### ISSUE THREAD

**Issue by witussk** [grafana/grafana#45803]
**Overview**
Hi, I am using SiteWise as a data source in AWS Managed Grafana, but it should apply to any other data source or Grafana version. 

I am trying to set visible thresholds on a gauge and graph based on a query, meaning that Grafana reads a specific measurement from AWS SiteWise, as it would read a value from Influx for example, and I would like to set thresholds based on this value. As it came to my attention, you cannot reference a variable in the thresholds menu. Then I came across the "Config from query result" transformation that let me set one threshold, but then I was not able to select another one, because it was unavailable.

My questions are:
Is this a bug that you cannot select more than one threshold from a query, or is it just supposed to be one? 
If it is just supposed to be one, is there another way how to show thresholds based on a query?

**What happened**:
I was unable to make another "Config from query result" transformation to achieve having multiple thresholds based on queries.

**What you expected to happen**:
Being able to have multiple thresholds by using this transformation method.

**How to reproduce it (as minimally and precisely as possible)**:
Have three different measurements, where two out of the three are the upper and lower thresholds and the third one is the Actual value. Attempt to set thresholds based on the two queries.

**Environment**:
- Grafana version: 8.2.5
- Data source type & version: AWS IoT SiteWise
- OS Grafana is installed on: AWS Managed Grafana
- User OS & Browser: Windows 11, Brave (Chrome)
- Grafana plugins: None, only ones included with AWS Managed Grafana
- Others: 

**Screenshot**
![image](https://user-images.githubusercontent.com/71400399/155378030-044912f9-e526-4857-a745-eeed7d434b70.png)

**Comment by torkelo** [comment]
The reason is simply that this transform is still pretty new, before adding to many features we wanted to understand that it was actually useful and what capabilities are needed. 

You data set looks to only have time, LSL, and quality fields , so what field would you use for thresholds 2?

**Comment by witussk** [comment]
> The reason is simply that this transform is still pretty new, before adding to many features we wanted to understand that it was actually useful and what capabilities are needed.
> 
> You data set looks to only have time, LSL, and quality fields , so what field would you use for thresholds 2?

I understand. So for different SiteWise measurements, separate queries are required for reading values. For example I have these:

![image](https://user-images.githubusercontent.com/71400399/155561765-4740fb36-4560-477f-a8a7-75adfb0c7b18.png)

And I would like to apply "LSL" and "SP" as thresholds to "Actual". That is the goal.

**Comment by atifali** [comment]
@witussk yes from what i understand there is no other way to do this as of now - addition to the config from query results transform would be a good long-term solution/feature to add - very useful indeed

**Comment by LukasRoskosch** [comment]
Hey, 
I´m stuck at the same point. I wanted to display a bar graph with two different thresholds so that it acts like a traffic light. The thresholds need to be dynamic and read out by a value queried from InfluxDB 2.1 with FLUX language. The transformation is fine with the first threshold, but adding another one does not work as shown by @witussk. 

I think it would be a great feature and I´m looking forward to that.

**Comment by vincent-gou** [comment]
Hi, Not a solution but i've encountered same kind of issue and you can define twice `thresold1` to different values from your request (tested in flux query).

![image](https://user-images.githubusercontent.com/46958513/182817716-004fc256-c379-4fd8-9a24-906d22705fae.png)

Maybe that solution may be usefull for you.

I agree that the ability to define 2 or more config query would be interesting.

Vincent

**Comment by hugosxm** [comment]
> The reason is simply that this transform is still pretty new, before adding to many features we wanted to understand that it was actually useful and what capabilities are needed.
> 
> You data set looks to only have time, LSL, and quality fields , so what field would you use for thresholds 2?

I feel that this feature is much needed in the case of the gauge graph. This way you can define all thresholds and maximum dynamically. For exemple in a monitoring dashboard you have a selector variable with the hostname of the machine you want to see. If you use the variable in every query you will have a dynamic dashboard for all your hosts, but the size of ram of hardisk may not be the same. Then you need a way to adjust thresholds...

**Comment by jrbenito** [comment]
hi,

It would be great if this transformation could grab data resulting from other transformations. In Prometheus for instance, we cannot grab more than one metric per query, hence, with some transformations we can have a table with fields representing min, max and some other thresholds. If config from result query could use data from this resulting instead of a fixed query, it would allow for Prometheus to provide multiple configuration points.

Thanks a lot.

**Comment by cmburns** [comment]
Hi,

is here any progress visible? It seems that we still can only choose one Transform.
I think it would be more if we can use variables (or query results) in gauge min, max, thershold, etc. fields directly instead of creating a transform.

**Comment by andreas-z** [comment]
+1
Useful feature in bar gauge panel to show CPU & memory usage with max set to the node's actual capacity (different VM sizes)

**Comment by jrbenito** [comment]
> I think it would be more easy if we can use variables (or query results) in gauge min, max, thershold, etc. fields directly instead of creating a transform.

I agree, however, min and max you already can source from query. Just leave the fields in auto on the gauge properties panel and have two fields in your query called min and max. It works automatically. The addition of more than one threshold would be great as well.

**Comment by cmburns** [comment]
@jrbenito: Min and Max inside query works - thanks!

To add a threshold using a query I tried the proposal by vincent-gou.
It is working using Thershold1 multiple times.

My question:
If I add e.g. 2 times Thershold1 - how can I set the color for each threshold?

**Comment by jrbenito** [comment]
I am happy that min/max worked for you!

> My question:
> If I add e.g. 2 times Thershold1 - how can I set the color for each threshold in the gauge?

That is also my doubt. I think we can't, but I hope to be wrong.

**Comment by jrbenito** [comment]
> Not a solution but i've encountered same kind of issue and you can define twice `thresold1` to different values from your request (tested in flux query).

That is a problem in Prometheus where you only get a single value per query. The ability to use config from query results over the result of previous transformation (or use multiple config from query results transformation) is key to be useful in Prometheus like queries. 

I wish I had the luxury of Flux sometimes! 

Thanks.

**Comment by vincent-gou** [comment]
> @jrbenito: Min and Max inside query works - thanks!
> 
> To add a threshold using a query I tried the proposal by **vincent-gou**. It is working using Thershold1 multiple times.
> 
> My question: If I add e.g. 2 times Thershold1 - how can I set the color for each threshold in the gauge?

For me not possible from my tests. 
But you can have different colors by data queries ( i mean not config query).

From my understandings:
- color value if set in config query has priority on color value defined in data query applied on by config query.
- color in config query only applies to color values  
- a different color values can be defined on a third data query for example 

![image](https://user-images.githubusercontent.com/46958513/213655500-633bf7df-e498-4480-bbcb-3b74efc9b9b3.png)

![image](https://user-images.githubusercontent.com/46958513/213655565-e4e3472d-5a48-44ee-8053-77dab59d8b08.png)

Solution would be to make thresold **color** customizable from config query and add extra thresold  settings like :
* thresold X + color
* thresold Y + color
* thresold Z + color ...

**Comment by vincent-gou** [comment]
> I wish I had the luxury of Flux sometimes!

Prometheurs has advantages that Flux doesn't have and vice versa...

**Comment by jrbenito** [comment]
> > I wish I had the luxury of Flux sometimes!
> 
> Prometheurs has advantages that Flux doesn't have and vice versa...

sure! We use both at the company, both serves different purposes here and both are awesome to work with.
My comment was joke because with flux I could put all things in a single query result and make config from query to it. But with Prometheus it is impossible because config from query applies to only one query and as mentioned, prometheus only give one value per query, so it is kind of useless to have config from query over a prometheus query.

**Comment by jon-ton** [comment]
> Solution would be to make thresold color customizable from config query and add extra thresold settings like :

    thresold X + color
    thresold Y + color
    thresold Z + color ...

Agree that this addition in the config transformation would be very helpful. This is one of the most commonly requested challenge for all of our users.

**Comment by AyoubHabra** [comment]
+1 
Very useful feature, especially for gauges and SLA dashboards.  
Also, if we can get multiple thresholds from config that would be really great .

**Comment by eproxus** [comment]
I think this is very much needed when working with Prometheus. Consider the following metrics:
* `db_pool_size` - the size of the pool
* `db_pool_max_overflow` - the max overflow workers allowed to be started
* `db_pool_workers` - the number of active workers
* `db_pool_workers_overflow` - the number of active overflow workers

If one wants to use the `_size` and `_max_overflow` for the config values _Max_ and _Threshold1_ respectively, it is currently impossible to do with the _Config from query results_ transformation. As soon as one of them is picked, it is impossible to use any of the other ones. Even if they all are merged using _Concatenate fields_ before, the transformation just shows _No options found_ when trying to select config query.

(If there is a way to transform these metrics so that they're possible to use with _Config from query results_ I'd really like to know. I feel like I've tried everything at this point).

**Comment by Darkextratoasty** [comment]
Is there still the limitation of just one _Config from query results_ transformation in Grafana? It seems like a bizarre limitation, especially considering questions about it go back as far as two years ago.

**Comment by fr3fou** [comment]
> > I think it would be more easy if we can use variables (or query results) in gauge min, max, thershold, etc. fields directly instead of creating a transform.
> 
> I agree, however, min and max you already can source from query. Just leave the fields in auto on the gauge properties panel and have two fields in your query called min and max. It works automatically. The addition of more than one threshold would be great as well.

Is this behavior documented anywhere? This is a very important feature imo and haven't been able to find any documentation regarding this

**Comment by tlifschitz** [comment]
Hello everyone, I think extending this functionality would indeed be very useful to many people!

I come to point out another use case example. Having a time series with multiple series (3 in this case), I have another query that returns a threshold for each series (called _vertical_desired_):

> ![image](https://github.com/grafana/grafana/assets/15240959/b167b923-be63-4d2c-b1a7-205d35bf4939)

If I choose 'All values' or  'All unique values' as the aggregation for the threshold, it does not work:

> ![image](https://github.com/grafana/grafana/assets/15240959/4d247c2a-6eeb-4112-9000-545d68243684) ![image](https://github.com/grafana/grafana/assets/15240959/706a1434-2600-4fe7-8657-97ae8e706c0f)

However if i chose an aggregation returning a single value it does display a threshold:

> ![image](https://github.com/grafana/grafana/assets/15240959/eea6a500-cd6c-4ff3-a140-3a3db85a9338) ![image](https://github.com/grafana/grafana/assets/15240959/a656c157-4bcd-4092-8a41-71737fcaa049)

It would be nice to make the threshold setting compatible with the 'All values' aggregation. Maybe even matching each threshold value by time series name? I understand that would be more sophisticated, but at least create all thresholds from the array values, for all series.

Can you take this into account @oscarkilhed ?  Is anyone pushing forward this effort?

**Comment by LarsStegman** [comment]
I spent some time on getting this working and I think it's ready for a PR (except missing tests)

<img width="1198" alt="image" src="https://github.com/grafana/grafana/assets/9332816/b7e23f29-0dce-48bb-81d3-6717d33354f2">

**Comment by omnistat** [comment]
> I spent some time on getting this working and I think it's ready for a PR (except missing tests)
> <img alt="image" width="1198" src="https://private-user-images.githubusercontent.com/9332816/307617989-b7e23f29-0dce-48bb-81d3-6717d33354f2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk0MTM3NTIsIm5iZiI6MTcxOTQxMzQ1MiwicGF0aCI6Ii85MzMyODE2LzMwNzYxNzk4OS1iN2UyM2YyOS0wZGNlLTQ4YmItODFkMy02NzE3ZDMzMzU0ZjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjZUMTQ1MDUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzAzNzI5NGMzMzgxNTZlNGY0ZGQyMjgxNDgwM2I0NzRkYTQ5NjI2YmMxYzA3MmUxMzAwYjYyZDBhYzU2NWVhOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.jTCIucnWB_PVdFi-UNjH4HyBVWFBxnAOBl6WujUh91A">

I'm sure this works great for data sources that can return multiple values from queries, but as far as I am aware, this is still impossible with Prometheus.

**Comment by douglasg14b** [comment]
What is the status on this item? Is this why I can't define thresholds from data like:

```csv
Text,Value,Color
Thresholds,10,orange
Thresholds,8,red
```

**Comment by Develer** [comment]
Hey, here is the [PR](https://github.com/grafana/grafana/pull/83366) that fixed the issue. Please, let us know if you have any further questions.
